### ISSUE TITLE
Improve usability

### ISSUE THREAD

**Issue by addyosmani** [addyosmani/getUserMedia.js#2]
I was hacking away at this this evening and discovered that at least for the demo, there are far too many configuration points needed for dimensions. If the shim can simplify this at all it should.

**Comment by franzenzenhofer** [comment]
given the different implementations of the different vendors i'm currently thinking about coding a lib ontop of this lib. the lib i have in mind would be simple
- getUserVideo(success, error)
- getUserAudio(success, error)

but the success callback would not (only) get the stream object, but a ready to use `<video>` or `<audio>` HTML element. 

the callback would be something like this

``` javascript
onSuccess = function(videoEl, stream) {
//do some stuff with the videoEl
//or go low level with the stream thingie
}
```

`<video>` and `<audio>` are pretty standardized over all modern browser and every dev. (by now) knows how to use them.

otherwise it will take ages until getUserMedia will get any mainstream adoption.

just my 2c on the usability topic

**Comment by addyosmani** [comment]
@franzenzenhofer If there are any other changes to the overall API that you would like to suggest I'd be happy to consider them for inclusion. At the moment whilst the library itself is very usable, I feel like better abstractions could help people get setup to use it in less time.

**Comment by franzenzenhofer** [comment]
started working on a simple  (cross browser but no flash fallback ) getUserVideo function, see https://github.com/franzenzenhofer/Sinne feel free to drop by (the video demo already works https://github.com/franzenzenhofer/Sinne/blob/master/demo/video.html )

**Comment by addyosmani** [comment]
@franzenzenhofer Awesome. I'll take a look! Our projects could work together. e.g eventually if yours was fleshed out enough, could drop the gUM logic in mine to consume yours and just use getUserMedia.js as a bridge to the fallback.

Will try to provide feedback.

**Comment by franzenzenhofer** [comment]
hi, i published https://github.com/franzenzenhofer/Sinne as V1, currently is uses the same logic a getUserMedia.js (rewritten in coffeescript), but yeah in the future it will make sense to use getUserMedia.noFallback.js directly. 

in the current implementation i don't see a way to use the flash fallback in an non-"OMG it's flash i have to implement a complete other logic for this"-way. 

i was thinking on how making a flash fallback feasible and non-disturbing, but the only way i could imagine was that we have to treat the flash-object as a kind of allow-bar.
- display the flash-thingie
- users clicks on "ok" for the video access
- flash thingie gets hidden
- flash thingie video stream gets funneled into a MediaStream (or whatever stream we need) object
- success callback gets called (with the stream)

as i happily gave up on flash in 2004,  i wonder if such a logic "flash thingie video stream -> MediaStream" is feasible?

**Comment by addyosmani** [comment]
Correct me if I've misinterpreted: Do you mean a way to display the Flash fallback in such a way that its completely (as in even from a UI perspective) transparent to the user what you're actually using behind the scenes? :) 

If this is correct, there are a few challenges with it:
- Chrome/Other browsers to the best of my knowledge don't support exposing the native notification/permissions bar for re-use. That said one could come up with something quite similar with CSS/HTML.
- There unfortunately isn't a way to override the Flash permissions dialog. People have found workarounds for this in the past, but because there were security implications of it those holes allowing it have been patched. One might be able to hide the permissions dialog button beneath something custom, but I wonder how well that might work.

From a usability perspective, because there's a noFallback version of the shim hosted in the repo at the moment, I'm wondering whether or not to implement the Flash fallback part as a plugin that gets optionally included in the build process. Beyond that I think it'll be down to making gUM more easy to use.

**Comment by franzenzenhofer** [comment]
hi

> Do you mean a way to display the Flash fallback in such a way that its completely (as in even from a UI perspective) transparent to the user what you're actually using behind the scenes? :)

no, not "completely", i mean a flow like this
- gerUserMedias.js getUserMedia(options, successCallback, errorCallback) gets called
- detects that there is no native getUserMedia()
- creates flash object
- display (gray rectangle) flash object - with permission dialogue
- user clicks "allow camera"
- hides (gray rectangle) flash object
- the flash (user webcam) video gets funneled into a MediaStream (or a MediaStream like) object 
- successCallback(stream) gets called

basically we only display the flash gray rectangle object for the permission dialogue - convert whatever flash gives us into a mediastream like object - call the successCallback

what i don't know, and what i want to find out (but i don't know who to ask), is, if this part "the flash (user webcam) video gets funneled into a MediaStream (or a MediaStream like) object" is in any way feasible?

> gUM more easy to use.
> working on it
